File: Interval/1893. Check if All the Integers in a Range Are Covered/solution2.py
class Solution:
    def isCovered(self, ranges: List[List[int]], left: int, right: int) -> bool:
        diff = [0]*52
        for l, r in ranges:
            diff[l] += 1
            diff[r+1] -= 1

        for i in range(1, len(diff)):
            diff[i] += diff[i-1]

        for i in range(left, right+1):
            if diff[i] == 0: return False
        return True

File: Tries/1268. Search Suggestions System/solution2.cpp
// great solution in terms of time and space
class Solution {
public:
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        vector<vector<string>> answer;
        int left = 0;
        int right = products.size()-1;
        sort(products.begin(), products.end());
        for(int i = 0; i < searchWord.size(); i++){
            char currChar = searchWord[i];
            while(left <= right && (products[left].length() <= i || products[left][i] != currChar)){
                left++;
            }
            while(right >= left && (products[right].length() <= i || products[right][i] != currChar)){
                right--;
            }
            int range = right - left + 1;
            answer.push_back({});
            for(int j = 0; j < min(3, range); j++){
                answer[answer.size()-1].push_back(products[left+j]);
            }
        }


        return answer;
    }
};